src/core/packet.lua

```diff
@@ -58,24 +60,36 @@ end
 
 -- Prepend data to the start of a packet.
 function prepend (p, ptr, len)
-   assert(p.length + len <= max_payload, "packet payload overflow")
-   C.memmove(p.data + len, p.data, p.length) -- Move the existing payload
+   shiftright(p, len)
    ffi.copy(p.data, ptr, len)                -- Fill the gap
-   p.length = p.length + len
    return p
 end
 
 -- Move packet data to the left. This shortens the packet by dropping
 -- the header bytes at the front.
 function shiftleft (p, bytes)
-   C.memmove(p.data, p.data+bytes, p.length-bytes)
+   assert(bytes >= 0 and bytes <= p.length)
+   p.data = p.data + bytes
+   p.headroom = p.headroom + bytes
```

Comment:
The invariant is that p.headroom + p.length <= max_payload.  This routine takes bytes away from length and gives them to headroom.